twisted/twisted
# [#7338] Add support for HTTP to HTTPS redirection on a TLS port.

This is still a proof of concept stage.

The ticket is now https://github.com/twisted/twisted/issues/7338

## Scope

This adds an HTTPS Site that handles TLS and HTTP to HTTPS redirection.

## Use case

You have a Twisted based application which runs the web admin dashboard on a custom port, ex 1234.

443 port is used by the user facing app.

Admin user remember that the admin dashboard is on port 1234.
The open a browser and type :  NAME.OF.THE.HOST:1234 
The browser (IE, Firefox, and Chrome as of today) will default to HTTP.
The connection fails as port 1234 is waiting for TLS handshake.
The admin user is panicked and calls  support :)


## To do

* Make mypy happy


## How to test


```
$ trial twisted.web.test.test_web
$ trial --without-module OpenSSL twisted.web.test.test_web
```

For manual test on Linux. Generate the server key+cert and then run the example in a terminal

``` 
$ openssl req -new -newkey rsa:2048 -days 3 -nodes -x509 \
        -keyout server-key.pem -out server-cert.pem
$ python docs/web/examples/https_http_redirect_server.py
```
In another terminal or a web browser try to access localhost over port 8443 with HTTP and HTTPS

```
    # See HTTP to HTTPS redirection.
    curl -v http://localhost:8443
    # See that HTTPS works
    curl -kv https://localhost:8443
```

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/490
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I ran additional checks listed at [Getting Your Patch Accepted](https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
